src/_articles/archive/zones.md

```diff
@@ -152,33 +154,51 @@ Code might execute in other nested child zones as well,
 but at a minimum it always runs in the root zone.
 
 
-## Handling asynchronous errors
+## Handling uncaught errors
 
 One of the most used features of zones is
-their ability to handle uncaught errors
-in asynchronously executing code.
-The concept is similar to a try-catch in synchronous code.
-
-An _uncaught error_ is often caused by some code using `throw`
-to raise an exception that is not handled by a `catch` statement.
-Another way to produce an uncaught error is
-to call `new Future.error()` or
-a Completer's `completeError()` method.
-
-Use the `onError` argument to `runZoned()` to
-install a _zoned error handler_—an asynchronous error handler
-that's invoked for every uncaught error in the zone.
-For example:
+their ability to catch and handle uncaught errors,
+both synchronous and asynchronous.
+An _uncaught error_ is often caused by code using `throw`
+to raise an exception without an accompanying `catch`
+statement to handle it.
```

Comment:
> Consider the asynchronous and synchronous clause. Wouldn't that be all errors? If so, exclude this clause.

Yeah I also just realized that sync and async is only true of `runZoneGuarded` anyway 

> Consider providing more context and chunking to get to an uncaught error. Explain it, then define it.

The text in the suggestion spends a lot of effort basically just explaining how error handling works, which is basic cs, not unique to Dart or zones, so doesn't need to be described. 

> Consider moving how zones help to the end. The prior sentences could build up to why it matters.

I don't really understand this. I'm in the school of main-point-first, details/why-following, because people are almost always skimming. Is this a principle we're using in practice? Can you show me other examples? This seems like a pretty obfuscating method for writing technical content.

Reply:
**reply:** I see your point. I tend to be in the "why should I care about this feature / item / topic" and like to lead folks a _little_ bit. If you just state what it is, I feel that conveys a finality to the text. It feels like a closed door to the face. I admit some of this comes from my background in ops where something tacitly known by devs doesn't get explained to others sometimes. I think that a little build up, a little context, makes the point more clear and better understood.

That said, I don't think that this is a hill for me to die on here and now. I think we can and should discuss it further later. If you're ok with it, let it ride. We can always come back to it.

**comment:** The original text does explain throw and catch, too, but "after the fact", which is why I recommended the switch.

Reply:
Nit: Thrown `Error`s do not *need* a corresponding `catch`. Errors shouldn't happen, so you shouldn't try to catch them. The person writing the `throw` should definitely never worry about that error being caught.

(You can have a global "catch all errors to prevent the app from crashing" error handler.)

Also, we don't throw "another exception" (the `_uncaught error_`). That's just a message printed when an uncaught error is reported.

Exceptions should always be caught, close to where they are thrown. But again, it's not the author writing the `throw` who writes the `catch`, because exceptions are really intended to be used across API boundaries.

So, the phrasing here, which links "when your code throws" to "needs related `catch`", conflates the responsibilities of two different authors, working on different sides of an API boundary.


It's tricky to say something general and decisive, because our recommendations are to not catch errors, but application developers really do like their apps to not crash, so they will catch errors and try to recover. The problem with errors is that it's not always clear whether you *can* recover.


I'd probably go back to what I wrote above, defining an uncaught error as:

> _Uncaught errors_ are `throw`s which reach the event loop without being caught by a `catch`.
> The can also come from a future completed with an error result, when nobody waits for that future.

That defines what the problem is, and gives it a name. The reader should agree to, and understand why, an uncaught error is a problem.
It doesn't say how to avoid the problem (well, `catch` and waiting for the future, but not in any detail).

> The current zone handles all uncaught errors. The default is to crash a native application.
> You can add your own custom _uncaught error handler_ to a new zone, handling them in your own way,
> to avoid uncaught errors crashing your program.

Reply:
**reply:** One important distinction: we can't use code elements as nouns or verbs. See the [GSG](https://developers.google.com/style/code-in-text#grammatical-treatment-of-code-elements) on that point.

I would also note my previous suggestion for some language for this point.